orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,568 @@
+                  candidate_node_args.find(cur->MutableInputDefs()[1]) != candidate_node_args.end());
+      NodeArg* arg_in_subgraph = nullptr;
+      NodeArg* arg_not_in_subgraph = nullptr;
+      if (candidate_node_args.find(cur->MutableInputDefs()[0]) != candidate_node_args.end()) {
+        arg_in_subgraph = cur->MutableInputDefs()[0];
+        arg_not_in_subgraph = cur->MutableInputDefs()[1];
+      } else if (candidate_node_args.find(cur->MutableInputDefs()[1]) != candidate_node_args.end()) {
+        arg_in_subgraph = cur->MutableInputDefs()[1];
+        arg_not_in_subgraph = cur->MutableInputDefs()[0];
+      }
+
+      // arg_in_subgraph is contained in candidate_node_args, so its shape must be [batch_size, seq_len, ...]
+      // Now only support cases of the two shapes are absolutely same or the other shape dim size is smaller by 2.
+      // For example, [batch_size, seqlen, hidden_size] and [batch_size, seqlen, hidden_size].
+      //              [batch_size, seqlen, hidden_size] and [seqlen].
+      // TODO: support other case such as:
+      //              [batch_size, seqlen, hidden_size] and [batch_size, 1, hidden_size]
+      if (arg_not_in_subgraph->Shape()->dim_size() <= arg_in_subgraph->Shape()->dim_size() - 2 ||
+          (arg_in_subgraph->Shape()->dim_size() == arg_not_in_subgraph->Shape()->dim_size() &&
+           arg_in_subgraph->Shape()->dim(0) == arg_not_in_subgraph->Shape()->dim(0) &&
+           arg_in_subgraph->Shape()->dim(1) == arg_not_in_subgraph->Shape()->dim(1))) {
+        candidate_node_args.insert(cur->MutableOutputDefs()[0]);
+        PushAllOutputNode(graph, to_visit, cur, visited);
+        // There are two possibilities here:
+        // 1. The size of arg_not_in_subgraph->Shape is smaller than arg_in_subgraph->Shape by 2,
+        //    do not need to add flatten pattern to arg_not_in_subgraph.
+        // 2. The size of arg_not_in_subgraph->Shape is same with arg_in_subgraph->Shape and the first
+        //    two dims value are exactly same, then there are also two possibilities:
+        //     <1>. The arg_not_in_subgraph is propagated from embedding_node (contained in candidate_node_args),
+        //          do not need to process it.
+        //     <2>. The arg_not_in_subgraph is not propagated from embedding_node (not contained in candidate_node_args),
+        //          need to add flatten pattern to arg_not_in_subgraph.
+        //     Here we just add cur node to candidate_input and process it (add flatten pattern to its input) after
+        //     the graph iteration, according to whether it's contained in candidate_node_args.
+        if (arg_in_subgraph->Shape()->dim_size() == arg_not_in_subgraph->Shape()->dim_size()) {
+          candidate_input.insert(cur);
+        }
+      } else {
+        LOG_DEBUG_INFO(logger, "PaddingElimination::Input shapes of node:" + cur->Name() + "are not compatible.");
+        candidate_output.insert(cur);
```

Comment:
curious why we call it candidate? will we change it later (adding or removing some in/from it) later? If not, shall we just call it subgraph_inputs/outputs?

Reply:
because here it's still not sure if it's input/output. Should wait to finish the iteration, candidate_inputs that are not contained in subgraph is real input and candidate_output that are contained in subgraph is real output.